### PR TITLE
Weapon system picking up

### LINKED ISSUE #11: (Group) Weapon System Picking UP
Create a class type for weapons (Alternatively, you can use scriptable objects)
- Allow for addition of different weapon types.

Create (or update) a script that will allow a character to pick up a weapon when they are within a certain distance.
- the weapon should be considered equipped
- the weapon on the map should be removed

Create (or update) a script that allows the player to drop a weapon they have.
- the weapon should be unequipped
- the weapon will then appear on the map

### PR DESCRIPTION
**Pull request by ghost**
Player can now Pickup and drop "Weapons"  based on a facing direction. right now that is Left or Right. A Pickup object is needed in the scene which holds a weapon scriptable object. this object is gotten by the player and "Equipped" in the player, when the player drops they creat a new pickup gameobject and assign it the data of the currently held wepaon

### CODE REVIEW

File: Assets/Scenes/DeveloperTemplate.unity
Comment (ExplosiveEggshells) [inline Assets/Scenes/DeveloperTemplate.unity:1]: Reminder: We ideally don't want to be making changes to the DeveloperTemplate scene (unless it is to fix any crashes that would occur from the new feature's changes).

Instead, we should make a duplicate of the DeveloperTemplate and test our new feature there. Try to revert these changes by checking out main, copying the original template, and then pasting it into your scene (After moving your stuff to a different scene, of course!)

Of course, I say this realizing that we haven't released a contributing doc that details this process, so it's technically my fault🤦‍♂️

File: Assets/Scripts/Controllers/CharacterController.cs
Comment (ExplosiveEggshells) [inline Assets/Scripts/Controllers/CharacterController.cs]: Consider commenting this, not exactly sure what `_weaponView` is.

```cs
[SerializeField, Tooltip("The root game object where new weapons will be parented to")] private GameObject _weaponView;`
```

File: Assets/Scripts/Controllers/CharacterController.cs
Comment (ExplosiveEggshells) [inline Assets/Scripts/Controllers/CharacterController.cs]: Logic for determining the facing of the character is useful 👍, let's keep this here in the controller when we isolate the pick up behavior.

File: Assets/Scripts/Controllers/CharacterController.cs
Comment (ExplosiveEggshells) [inline Assets/Scripts/Controllers/CharacterController.cs]: Keep this in the character controller when we isolate the pickup behavior.

File: Assets/Scripts/Controllers/CharacterController.cs
Comment (ExplosiveEggshells) [inline Assets/Scripts/Controllers/CharacterController.cs]: We'll definitely want to keep some Interaction input processing in the controller. However, we should generalize this code into something that abstractly tries to _inform_ an interactable object that it has been interacted with, and allow the interactable to then decide what to do in response. This is generally a much more scalable approach.

Again, this is something I'd be happy to help you with. It will involve creating a `IInteractible` interface and some slight changes to your pick up code.

File: Assets/Scripts/Controllers/CharacterController.cs
Comment (ExplosiveEggshells) [inline Assets/Scripts/Controllers/CharacterController.cs]: This is the major portion that we'll want to pull out into a different "module" to extend the character's behavior, rather than adding it to the controller. 

You'd likely want to create a `CharacterItemManager.cs`, whose sole responsibility is to handle picking up, dropping, and propagating events towards the currently equipped item. Again, we'll want to wait until #33 is merged in to do this, so we can discuss more after that.

File: Assets/Scripts/PickupObject.cs
Comment (ExplosiveEggshells) [inline Assets/Scripts/PickupObject.cs]: Seems to be an unnecessary using statement? Consider removal if not needed.

File: Assets/Scripts/PickupObject.cs
Comment (ExplosiveEggshells) [inline Assets/Scripts/PickupObject.cs]: We should probably also include a warning message to indicate to designers that the weapon is not configured properly. Further, disabling the object might be a more appropriate response than destroying it:

```cs
...
gameObject.SetActive(false);
Debug.LogWarning($"PickupObject::Start - Weapon data is null, disabling self.");
...
```

File: Assets/Scripts/PickupObject.cs
Comment (ExplosiveEggshells) [inline Assets/Scripts/PickupObject.cs]: Nit: Correct spelling to "InitializeObject"

File: Assets/Scripts/Weapon.cs
Comment (ExplosiveEggshells) [inline Assets/Scripts/Weapon.cs:1]: This works for now, but it's worth noting that many of our equippable items are not weapons, and may not use a concept of damage.

In the future, we'll likely make an issue to introduce a new member to this inheritence:
`EquippableItem -> Weapon -> Pistol`

Where EquippableItems are a more generic form of items that can be held by the player.

File: Assets/Scripts/Weapon.cs
Comment (ExplosiveEggshells) [inline Assets/Scripts/Weapon.cs]: Per convention: `_Worldsprite` -> `_worldSprite`

**Review by ExplosiveEggshells — DISMISSED** [review]
Hey Adam, thanks for working on this, check out the comments for more specific details, but here's an overview:

Firstly, #33 is introducing some changes that is going to make the `CharacterController.cs` file _solely_ responsible for just processing input data and making it publicly available to other files. We will then pull out any other functionality into separate modules, for example.

TL;DR - We're moving away from shoving all of the functionality into the controller, so we'd also like to see this PR conform to that. This will involve pulling the logic in the controller out to something like `CharacterItemHandler.cs`.

We'll need to wait for #33 to merge to do so, but this is worth considering for now. I'm more than happy to set up a time to help you with this if you'd like (either scheduled, or during our Friday workshop).

As always, appreciate the work!

File: Assets/Scripts/PickupObject.cs
Comment (ghost) [reply Assets/Scripts/PickupObject.cs]: I have no idea what i did that made VS code or VS:C add that, it has been removed

File: Assets/Scripts/PickupObject.cs
Comment (ghost) [reply Assets/Scripts/PickupObject.cs]: Noted, logic changed accordingly.
To better pointout what object is throwing the warning I have happened the gameobject name into the message, is this a good idea? or am I off base?
Debug.LogWarning($"PickupObject::Start At Object " + gameObject.name + " - Weapon data is null, disabling self.");

Comment (ExplosiveEggshells) [reply Assets/Scripts/PickupObject.cs]: Yes! It is usually pretty good to include the object name as well.

FYI- you don't need to use string concat, and can instead utilize an interpolated string. [I've written about them here.](https://library.niugame.dev/guides/unity-guides/cs-language-features/#interpolated-strings)

```cs
Debug.LogWarning($"PickupObject::Start on Object {gameObject.name} - Weapon data is null, disabling self.");
```

**Comment by ghost** [comment]
I Have Addressed All but the item system expansion comment., and I would like to talk to you at some point this week and have you look at the interface system I have working, I can talk almost anytime today/tonight, If not today I can DM you my availability this week at your convince.

**Comment by ghost** [comment]
nothing is diffrent from when we looked at it, i forgot to call this after i looked through the code again

**Review by Migs — APPROVED** [review]

**Pull request merged**
